### PR TITLE
Add HTTP Request Timeout

### LINKED ISSUE #958: AppInsights blocks Node.JS process from exiting for hours
Even very simple telemetry submission code is preventing NodeJS process from exiting. Sometimes for minutes, sometimes for hours. The problem is especially noticeable on Mac OS X, where the delays are always at least a minutes.

This is being used in Azure DevOps extension, that adds build task into the pipeline. Adding minutes or hours to the build is really a big problem.

Removing AppInisghts makes the code run under a second.

```javascript
  try {
      addBuildVariableTelemetry();           
      const totalTimeTaken = Math.round((Date.now() - startTime) / 1000); // get time in seconds
      addMetric("ExecutionTime", totalTimeTaken);             

      const telemetryClient = new appInsights.TelemetryClient(TELEMETRY_KEY);
      
      /* None of this helps 
      telemetryClient.config.maxBatchIntervalMs = 0;
      telemetryClient.config.maxBatchSize = 0;        
      telemetryClient.config.httpsAgent = new https.Agent();
      telemetryClient.config.enableAutoCollectPerformance = false ; 
      telemetryClient.config.enableUseAsyncHooks = false;
      */
      
      telemetryClient.trackEvent({ name: taskName, properties: EventProperties as any, measurements: Metrics as any });
      if (Traces.length > 0) {
          telemetryClient.trackTrace({ message: Traces.join("\n"), severity: appInsights.Contracts.SeverityLevel.Information, properties: EventProperties as any});
      }

      telemetryClient.flush();        
      tl.debug("Telemetry submitted");
      hasPublishedFinalTelemetry = true;
  }
  catch (error) {
      console.log("##vso[task.LogIssue type=warning;]Telemetry wrapper:", error); // don't let telemetry crash us
  }
```

### PR DESCRIPTION
**Pull request by JacksonWeber**
Fixes #958 by adding a default HTTP request timeout of 20 seconds. Includes associated unit test.

### CODE REVIEW

File: Library/Sender.ts [hunk 1]
```diff
@@ -29,6 +29,7 @@ class Sender {
     public static CLEANUP_TIMEOUT = 60 * 60 * 1000; // 1 hour
     public static FILE_RETEMPTION_PERIOD = 7 * 24 * 60 * 60 * 1000; // 7 days
     public static TEMPDIR_PREFIX: string = "appInsights-node";
+    public static HTTP_TIMEOUT: number = 20000; // 20 seconds
```
Comment (hectorhdzg) [inline Library/Sender.ts:32]: We need this in CorrelationIdManager and AzureMetaData classes as well

File: Library/Sender.ts [hunk 2]
```diff
@@ -312,6 +321,11 @@ class Sender {
 
                     if (typeof callback === "function") {
                         if (error) {
+                            // If the error type is a timeout we want to provide more meaningful output
```
Comment (hectorhdzg) [inline Library/Sender.ts:324]: What is original error name and message here?

Comment (JacksonWeber) [reply Library/Sender.ts:324]: The original is: 
```
{ message: 'socket hang up', name: 'Error'" }
```

File: Library/CorrelationIdManager.ts [hunk 3]
```diff
@@ -7,6 +7,7 @@ class CorrelationIdManager {
     private static _handle: NodeJS.Timer;
     public static correlationIdPrefix = "cid-v1:";
     public static w3cEnabled = true;
+    public static HTTP_TIMEOUT: number = 20000; // 20 seconds
```
Comment (hectorhdzg) [inline Library/CorrelationIdManager.ts:10]: You can leave this as 2.5 seconds as well

File: Library/CorrelationIdManager.ts [hunk 4]
```diff
@@ -79,9 +81,17 @@ class CorrelationIdManager {
                 }
             }, true, false);
             if (req) {
+                req.setTimeout(CorrelationIdManager.HTTP_TIMEOUT, () => {
```
Comment (hectorhdzg) [inline Library/CorrelationIdManager.ts:84]: You are missing the test for this one

**Review by hectorhdzg — APPROVED** [review]
LGTM

**Comment by PavelBansky** [comment]
Very nice. When it will be available for download? Which version on the npm package?

**Comment by hectorhdzg** [comment]
@PavelBansky we will publish this as part of next release, end of this month.

**Pull request merged**